docs/syntax/syntax.md

```diff
@@ -205,95 +205,429 @@ Is below, but it won't be parsed into the document.
 
 (syntax/referencing)=
 
-## Markdown Links and Referencing
+## Links and Referencing
 
-Markdown links are of the form: `[text](link)`.
+CommonMark links come in three forms ([see the spec](https://spec.commonmark.org/0.30/#links)):
 
-If you set the configuration `myst_all_links_external = True` (`False` by default),
-then all links will be treated simply as "external" links.
-For example, in HTML outputs, `[text](link)` will be rendered as `<a href="link">text</a>`.
+*Autolinks* are [URIs][uri] surrounded by `<` and `>`:
 
-Otherwise, links will only be treated as "external" links if they are prefixed with a scheme,
-configured with `myst_url_schemes` (by default, `http`, `https`, `ftp`, or `mailto`).
-For example, `[example.com](https://example.com)` becomes [example.com](https://example.com).
+```md
+<scheme:path?query#fragment>
+```
 
-:::{note}
-The `text` will be parsed as nested Markdown, for example `[here's some *emphasised text*](https://example.com)` will be parsed as [here's some *emphasised text*](https://example.com).
+*Inline links* allow for optional explicit text and titles (in HTML titles are rendered as tooltips):
+
+```md
+[Explicit *Markdown* text](destination "optional explicit title")
+```
+
+or, if the destination contains spaces,
+
+```md
+[text](<a destination>)
+```
+
+*Reference links* define the destination separately in the document, and can be used multiple times:
+
+```md
+[Explicit *Markdown* text][label]
+
+[label]: destination "optional explicit title"
+```
+
+MyST, supports the following destination types:
+
+|     Link Type      |            Auto            |          Inline           | Single Page[^sp] |
+| :----------------- | :------------------------- | :------------------------ | :--------------: |
+| External URL       | `<https://example.com>`    | `[](https://example.com)` |        ✅         |
+| Local file path    | `<path:file.txt>`          | `[](file.txt)`            |        ❌         |
+| Project document   | `<project:file.md>`        | `[](file.md)`             |        ❌         |
+| Local target       | `<project:.#target>`       | `[](.#target)`            |        ✅         |
```

Comment:
What's the benefit of making the target explicitly local via `.#`? Is that just in case you have duplicated target IDs in different documents of a project?  We should document this behavior explicitly, e.g. a section like `### Restrict target search to the local document`

Reply:
Is the logic what @choldgraf specified, it **only** looks to the local document?

Do other links still look to all implicit/explicit references in the local document first regardless?

Reply:
@rowanc1

So firstly to note, there are no implicit targets stored by sphinx, only a unique set of target names is stored, per `domain/object_type`, e.g.

```yaml
std:
  label:
    api/directive:
      docname: api/reference
      id: api-directive
      text: Directive and role processing
```

If sphinx encounters a `domain/object_type/target` that is already in the "database", then it will simply emit a warning.
But essentially, we don't have that information at resolution time.

With the MyST heading anchors extension,  which the user has to explicitly activate with e.g. `myst_heading_anchors = 2`, we create an exception to this rule, by explicitly storing all anchors by `docname/domain/object_type/target`, i.e. this is the only time when we might have duplicate target names for the same `domain/object_type`

For e.g. `[](#target)` then:

1. If heading anchors are not activated, it will simply look "project-wide" to find matches.
    If multiple matches are found, then a warning will be emitted, and (arbitrarily) the first match will be selected, e.g.
    ```
    WARNING: Multiple targets found for '*:*:target': 'std:label:target','std:doc:target' [myst.xref_ambiguous]
    ```
2. If heading anchors are activated, it will look for both matches in the anchors "database" (for that docname) and matches in the "project" database (no docname filter).
    If both a heading anchor match and other matches are present, then it will emit a warning and select that e.g.
    ```
    <src>/test.md:6: WARNING: 'target' anchor superseding other matches: 'std:label:target' [myst.xref_anchor]
    ``` 

The difference with using `[](other.md#target)`, is that (a) project matches will be filtered by that docname and (b) heading anchors will be matched by that docname.

If you use `[](.#target)` and e.g. the file that the reference is in is `test.md`, then this is the same as doing `[](test.md#target)`

Does that make sense?

A key design consideration here, is that I didn't want any "silent" selection between multiple matches, i.e. there is always a warning.

But then the user can choose to ignore/suppress this warning, by type e.g.

```python
suppress_warnings = ["myst.xref_ambiguous", "myst.xref_anchor"]
```